### PR TITLE
V3 android cmake support

### PR DESCRIPTION
**Pull request by newnon**
CMake build improvements & android CMake support 
add to graddle.properties
PROP_NDK_MODE=cmake 
replace files in folder external/android-specific 
[android-specific.zip](https://github.com/cocos2d/cocos2d-x/files/1217221/android-specific.zip)

last PR https://github.com/cocos2d/cocos2d-x/pull/17100

pull request for external
https://github.com/cocos2d/cocos2d-x-3rd-party-libs-bin/pull/300

### CODE REVIEW

**Comment by minggo** [comment]
Looks good, so i have to merge cocos2d/cocos2d-x-3rd-party-libs-bin#300 to test it right? Could you please show me the steps to use cmake in gradle, thanks?

**Comment by newnon** [comment]
1. ensure that you have installed NDK & CMake from android studio package manager
2. update externals(just few changes in android_specific folder)
3. add to graddle.properties PROP_NDK_MODE=cmake(without it it will use ndkbuild)
4. make clean build

**Comment by minggo** [comment]
thanks, i have CMake installed in system(Mac OS), should the CMake be installed from android studio package manger again?

And in which path should i run `make clean build`?

**Comment by newnon** [comment]
Yes you shod install it from package manager(it will be installed inside SDK folder) if google does not change this, android studio does not use system CMake.

**Comment by crazyhappygame** [comment]
Guys,
When we may expect cmake to see in v3 branch?

**Comment by drelaptop** [comment]
Sorry for a long response, I really appreciate this patch. @newnon 

Following your steps, I have tested it using latest code, it occurred a path error ,and then I fixed it in [this commit](https://github.com/drelaptop/cocos2d-x/commit/386b718e618aa4af33cf9286092e64ce1f71219c). finally `cpp-tests` and `cpp-empty-test` build works well both __cmake build__ and __ndkBuild__.

Suggestion:

1. to merge the latest code into your branch to solve conflicts
1. fix path error, from `proj.android/jni/main.cpp` to `proj.android/app/jni/main.cpp`
1. add `PROP_NDK_MODE=cmake` to `gradle.properties` for all `proj.android`, and do some comments in detail.
comments like:
```
# android cmake support
# uncomment it, native code will build by cmake
# keep comment, native code will build by ndkBuild
#PROP_NDK_MODE=cmake
```
4. do further test, and tell what done and what undone, we should test in detail as possible as we can before merge. I will do it with you.

**Comment by drelaptop** [comment]
@crazyhappygame `cmake support android` will be done in 3.17, this is I wish. 

support other platforms is on schedule, but I couldn't give you the deadline,  we can do it together to make it done earlier.

**Comment by newnon** [comment]
OK I will try to do it as soon as possible.

**Comment by drelaptop** [comment]
`CMake support for js-tests android project` occurred errors, when I try to run js-tests using CMake build, please have a test after you adding cmake config to js-tests, and try to fix this. 

you know we have to make sure all android project support cmake well, the tests project, the template project, even the lib project. really a lot of work to do for us.

**Comment by crazyhappygame** [comment]
@newnon would it be possible with cmake to build cocos2d-x once and use build output libraries in other games? I would like to save compilation time for my android games.

**Comment by drelaptop** [comment]
Any progress about cmake ? @newnon 

You can have a try to use the output libraries in your game, hope for your feedback @crazyhappygame

**Comment by newnon** [comment]
@drelaptop cleared version rebased on top of current v3

**Comment by newnon** [comment]
@crazyhappygame unfortunately i haven't try this options, but it is possible.

File: cocos/network/CMakeLists.txt [hunk 1]
```diff
@@ -10,34 +10,37 @@ else()
 endif()
 
 if(ANDROID)
-   set(COCOS_NETWORK_PLATFORM_SRC
-        network/CCDownloader-android.cpp
-        )
-elseif(MACOSX)
-   set(COCOS_NETWORK_PLATFORM_SRC
-        network/CCDownloader-apple.mm
-        )
```
Comment (drelaptop) [inline cocos/network/CMakeLists.txt]: error occur when you delete this line, compile error on mac:
```
[ 82%] Linking CXX executable ../../bin/cpp_tests.app/Contents/MacOS/cpp_tests
Undefined symbols for architecture x86_64:
  "cocos2d::network::DownloaderApple::DownloaderApple(cocos2d::network::DownloaderHints const&)", referenced from:
      cocos2d::network::Downloader::Downloader(cocos2d::network::DownloaderHints const&) in libcocos2dInternal.a(CCDownloader.cpp.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
  
you can merge this to fix https://github.com/newnon/cocos2d-x/pull/5

**Review by drelaptop — CHANGES_REQUESTED** [review]
when change the cmake files of Android, I suggest keep other platform logic same. If not, do test to confirm the logic right.

**Comment by drelaptop** [comment]
[This commit](https://github.com/newnon/cocos2d-x/pull/5/commits/98d9ee0d6ae06341cf24319d1b45cb13ae1ed2d5) fix js-tests build error on mac, and build success, but run failed, only a black box appear. please help to do some check if you have free time.

#18545 have been fixed too, run failed is strange, please give some instructions about black box if you have time recently. @dumganhar

**Comment by drelaptop** [comment]
Sorry, some config difference exist on js-tests, build right and run success, App path is different from others, see `bin` files tree:
```sh
├── cpp_empty_test.app
│   └── Contents
├── cpp_tests.app
│   └── Contents
├── js-tests
│   ├── js-tests
│   ├── main.js
│   ├── project.json
│   ├── res
│   ├── script
│   └── src
├── lua-tests.app
│   └── Contents
└── lua_empty_test.app
    └── Contents
```
`bin/js-tests/js-tests` is the `js-tests` project App, which is different from `cpp-tests.app`
  
  
__Now, this branch on macOS(not include iOS) platform, cocos2d-x cmake build success, all tests project run success too. `cmake .. -G ”Xcode“` generate Xcode macOS project success, tests project except `js-tests` run in Xcode well.__
> js-tests project run in Xcode need to be look into
> project by `cocos new` need to be test

**Comment by drelaptop** [comment]
#18646 have been merged, this PR can be closed directly, thank you @newnon

**Comment by minggo** [comment]
@drelaptop please write a doc for it and ask developers to test it. Thanks.

**Comment by drelaptop** [comment]
yeah, I should do that, a sample guide have been added in the http://discuss.cocos2d-x.org/t/in-progress-cmake-support-all-platform-for-cocos2d-x/41164/13

I will add cmake android guide later

**Pull request closed without merging**